File: aoc-2023/src/main/java/com/capital7software/aoc/aoc2023/days/Day11.java
package com.capital7software.aoc.aoc2023.days;

import com.capital7software.aoc.lib.AdventOfCodeSolution;
import com.capital7software.aoc.lib.analysis.CosmicExpansion;
import java.time.Instant;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * --- Day 11: Cosmic Expansion ---<br><br>
 * You continue following signs for "Hot Springs" and eventually come across an observatory.
 * The Elf within turns out to be a researcher studying cosmic expansion using the giant
 * telescope here.
 *
 * <p><br>
 * He doesn't know anything about the missing machine parts; he's only visiting for this research
 * project. However, he confirms that the hot springs are the next-closest area likely to have
 * people; he'll even take you straight there once he's done with today's observation analysis.
 *
 * <p><br>
 * Maybe you can help him with the analysis to speed things up?
 *
 * <p><br>
 * The researcher has collected a bunch of data and compiled the data into a single giant image
 * (your puzzle input). The image includes empty space (.) and galaxies (#). For example:
 *
 * <p><br>
 * <code>
 * ...#......<br>
 * .......#..<br>
 * #.........<br>
 * ..........<br>
 * ......#...<br>
 * .#........<br>
 * .........#<br>
 * ..........<br>
 * .......#..<br>
 * #...#.....<br>
 * </code>
 *
 * <p><br>
 * The researcher is trying to figure out the sum of the lengths of the shortest
 * path between every pair of galaxies. However, there's a catch: the universe
 * expanded in the time it took the light from those galaxies to reach the observatory.
 *
 * <p><br>
 * Due to something involving gravitational effects, only some space expands.
 * In fact, the result is that any rows or columns that contain no galaxies should all
 * actually be twice as big.
 *
 * <p><br>
 * In the above example, three columns and two rows contain no galaxies:
 *
 * <p><br>
 * <code>
 * v  v  v<br>
 * ...#......<br>
 * .......#..<br>
 * #.........<br>
 * &#62;..........&#60;<br>
 * ......#...<br>
 * .#........<br>
 * .........#<br>
 * &#62;..........&#60;<br>
 * .......#..<br>
 * #...#.....<br>
 * ^  ^  ^<br>
 * </code>
 *
 * <p><br>
 * These rows and columns need to be twice as big; the result of cosmic expansion therefore
 * looks like this:
 *
 * <p><br>
 * <code>
 * ....#........<br>
 * .........#...<br>
 * #............<br>
 * .............<br>
 * .............<br>
 * ........#....<br>
 * .#...........<br>
 * ............#<br>
 * .............<br>
 * .............<br>
 * .........#...<br>
 * #....#.......<br>
 * </code>
 *
 * <p><br>
 * Equipped with this expanded universe, the shortest path between every pair of galaxies
 * can be found. It can help to assign every galaxy a unique number:
 *
 * <p><br>
 * <code>
 * ....1........<br>
 * .........2...<br>
 * 3............<br>
 * .............<br>
 * .............<br>
 * ........4....<br>
 * .5...........<br>
 * ............6<br>
 * .............<br>
 * .............<br>
 * .........7...<br>
 * 8....9.......<br>
 * </code>
 *
 * <p><br>
 * In these 9 galaxies, there are 36 pairs. Only count each pair once; order within the pair
 * doesn't matter. For each pair, find any shortest path between the two galaxies using only
 * steps that move up, down, left, or right exactly one . or # at a time. (The shortest path
 * between two galaxies is allowed to pass through another galaxy.)
 *
 * <p><br>
 * For example, here is one of the shortest paths between galaxies 5 and 9:
 *
 * <p><br>
 * <code>
 * ....1........<br>
 * .........2...<br>
 * 3............<br>
 * .............<br>
 * .............<br>
 * ........4....<br>
 * .5...........<br>
 * .##.........6<br>
 * ..##.........<br>
 * ...##........<br>
 * ....##...7...<br>
 * 8....9.......<br>
 * </code>
 *
 * <p><br>
 * This path has length 9 because it takes a minimum of nine steps to get from galaxy 5 to galaxy
 * 9 (the eight locations marked # plus the step onto galaxy 9 itself). Here are some other
 * examples of shortest path lengths:
 *
 * <p><br>
 * <code>
 * Between galaxy 1 and galaxy 7: 15<br>
 * Between galaxy 3 and galaxy 6: 17<br>
 * Between galaxy 8 and galaxy 9: 5<br>
 * </code>
 *
 * <p><br>
 * In this example, after expanding the universe, the sum of the shortest path between all
 * 36 pairs of galaxies is 374.
 *
 * <p><br>
 * Expand the universe, then find the length of the shortest path between every pair of galaxies.
 * What is the sum of these lengths?
 *
 * <p><br>
 * Your puzzle answer was 9686930.
 *
 * <p><br>
 * --- Part Two ---<br><br>
 * The galaxies are much older (and thus much farther apart) than the researcher initially
 * estimated.
 *
 * <p><br>
 * Now, instead of the expansion you did before, make each empty row or column one million
 * times larger. That is, each empty row should be replaced with 1000000 empty rows, and each
 * empty column should be replaced with 1000000 empty columns.
 *
 * <p><br>
 * (In the example above, if each empty row or column were merely 10 times larger, the sum
 * of the shortest paths between every pair of galaxies would be 1030. If each empty row or
 * column were merely 100 times larger, the sum of the shortest paths between every pair of
 * galaxies would be 8410. However, your universe will need to expand far beyond these values.)
 *
 * <p><br>
 * Starting with the same initial image, expand the universe according to these new rules,
 * then find the length of the shortest path between every pair of galaxies. What is the sum
 * of these lengths?
 *
 * <p><br>
 * Your puzzle answer was 630728425490.
 */
public class Day11 implements AdventOfCodeSolution {
  private static final Logger LOGGER = LoggerFactory.getLogger(Day11.class);

  /**
   * Instantiates this Solution instance.
   */
  public Day11() {

  }

  @Override
  public String getDefaultInputFilename() {
    return "inputs/input_day_11-01.txt";
  }

  @Override
  public void runPart1(List<String> input) {
    var start = Instant.now();
    int gapFactor = 1;
    runParts(input, start, gapFactor);
  }

  @Override
  public void runPart2(List<String> input) {
    var start = Instant.now();
    int gapFactor = 999_999;
    runParts(input, start, gapFactor);
  }

  private void runParts(List<String> input, Instant start, int gapFactor) {
    var answer = sumOfAllPairsShortestPath(input, gapFactor, false);
    var end = Instant.now();

    LOGGER.info("With a Gap Factor of {} the sum of all Galaxy pairs shortest path is: {}",
                gapFactor, answer);
    logTimings(LOGGER, start, end);

    start = Instant.now();
    answer = sumOfAllPairsShortestPath(input, gapFactor, true);
    end = Instant.now();

    LOGGER.info(
        "With a Gap Factor of {} the sum of optimized all Galaxy pairs shortest path is: {}",
        gapFactor, answer
    );
    logTimings(LOGGER, start, end);
  }

  /**
   * Calculates and returns the sum of the shortest paths between all pairs of Galaxies.
   *
   * @param input     The List of Strings that contain the Galaxies to parse.
   * @param gapFactor The number of additional spaces to count for each gap encountered.
   * @param optimized If true, uses a time and space efficient algorithm to calculate the
   *                  distances.
   * @return The sum of the shortest paths between all pairs of Galaxies!
   */
  public long sumOfAllPairsShortestPath(List<String> input, int gapFactor, boolean optimized) {
    var cosmos = CosmicExpansion.buildUniverse(input);

    return cosmos.sumOfAllPairsShortestPath(gapFactor, optimized);
  }
}

File: aoc-2017/src/main/kotlin/com/capital7software/aoc/aoc2017aoc/days/Day11.kt
package com.capital7software.aoc.aoc2017aoc.days

import com.capital7software.aoc.lib.AdventOfCodeSolution
import com.capital7software.aoc.lib.grid.HexEd
import java.time.Instant
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * **--- Day 11: Hex Ed ---**
 *
 * Crossing the bridge, you've barely reached the other side of the stream when a
 * program comes up to you, clearly in distress. "It's my child process," she says,
 * "he's gotten lost in an infinite grid!"
 *
 * Fortunately for her, you have plenty of experience with infinite grids.
 *
 * Unfortunately for you, it's a hex grid.
 *
 * The hexagons ("hexes") in this grid are aligned such that adjacent hexes can be
 * found to the north, northeast, southeast, south, southwest, and northwest:
 *
 * ```
 *   \ n  /
 * nw +--+ ne
 *   /    \
 * -+      +-
 *   \    /
 * sw +--+ se
 *   / s  \
 * ```
 *
 * You have the path the child process took. Starting where he started, you need to
 * determine the fewest number of steps required to reach him. (A "step" means to
 * move from the hex you are in to any adjacent hex.)
 *
 * For example:
 *
 * - ne,ne,ne is 3 steps away.
 * - ne,ne,sw,sw is 0 steps away (back where you started).
 * - ne,ne,s,s is 2 steps away (se,se).
 * - se,sw,se,sw,sw is 3 steps away (s,s,sw).
 *
 * Your puzzle answer was **705**.
 *
 * **--- Part Two ---**
 *
 * **How many steps away** is the **furthest** he ever got from his starting position?
 *
 * Your puzzle answer was **1469**.
 */
class Day11 : AdventOfCodeSolution {
  companion object {
    private val log: Logger = LoggerFactory.getLogger(Day11::class.java)
  }

  override fun getDefaultInputFilename(): String = "inputs/input_day_11-01.txt"

  override fun runPart1(input: MutableList<String>) {
    val start = Instant.now()
    val answer = fewestStepsToReachChild(input.first())
    val end = Instant.now()

    log.info("$answer is the fewest steps to reach the child process in the hex grid!")
    logTimings(log, start, end)
  }

  override fun runPart2(input: MutableList<String>) {
    val start = Instant.now()
    val answer = furthestDistanceReached(input.first())
    val end = Instant.now()

    log.info("$answer is the furthest the child process got from the starting position!")
    logTimings(log, start, end)
  }

  /**
   * Returns the fewest number of steps to reach the child process.
   *
   * @param input The [String] path that the child process took.
   * @return The fewest number of steps to reach the child process.
   */
  fun fewestStepsToReachChild(input: String): Int {
    val instance = HexEd(input)
    return instance.stepsAndDistance.first
  }

  /**
   * Returns the furthest distance that the child process got from his starting position.
   *
   * @param input The [String] path that the child process took.
   * @return The furthest distance that the child process got from his starting position.
   */
  fun furthestDistanceReached(input: String): Int {
    val instance = HexEd(input)
    return instance.stepsAndDistance.second
  }
}
